### PR TITLE
I added new physically based present throwing

### PR DESCRIPTION
**Pull request by patm1987**
We don't have a physics engine in the game yet (and with our current scoring system, I don't think we want one?).

Instead I've moved from a cubic b-spline to a naive computation of initial y velocity given the distance to the target and how fast we want the present to travel in the x/z plane. This is hard to tune for our test level, but I think when we have better targets in place this will feel more natural?

### CODE REVIEW

File: static/scenes/railroad/js/present.js [hunk 1]
```diff
@@ -28,11 +25,27 @@ class Present {
...
-        this.totalFlightTime = .25;
+
+        // start of the throw
+        this.startPosition = new THREE.Vector3();
+
+        // end of the throw
+        this.endPosition = new THREE.Vector3();
+
+        // how long it's going to take to complete the throw
+        this.lengthOfThrow = 0;
```
Comment (Jezzamonn) [inline static/scenes/railroad/js/present.js]: suggestion: lengthOfThrow -> durationOfThrow, to make it's clear it's time instead of distance.

**Review by Jezzamonn — APPROVED** [review]
Nice!

**Comment by Jezzamonn** [comment]
Ooh, ok I tried this out. When you throw things far away, the presents get launched way too high IMO.

Curious as to what it would look like if we faked gravity and just had some variable height based on distance instead. What do you think?

**Comment by Jezzamonn** [comment]
And to clarify, it looks good when you throw presents nearby

**Comment by Jezzamonn** [comment]
Actually, the presents also take a long time to get to their destination... maybe we clamp the total time to 0.5 seconds or something. Then the height will end up being a little more normal I think.

**Comment by patm1987** [comment]
WRT the presents going too high and the duration: all of it is faked since I'm not entirely sure what our scale is. The physics are all faked though, so I can get clever capping both (.5 will be painfully fast for some of the distances involved though).

File: static/scenes/railroad/js/present.js [hunk 2]
```diff
@@ -5,11 +5,9 @@ const loader = new THREE.OBJLoader();
 
 let loadedObj;
 
-// the distance to the target scaled by this number is how high we go
-const midpointOffset = .25;
-
-// a value from 0 to 1 that defines the peak height
-const midpointT = .25;
+const gravity = -20; // acceleration along y
+const linearThrowSpeed = 20; // velocity in the x/z plane
```
Comment (Jezzamonn) [inline static/scenes/railroad/js/present.js:9]: I think as you noticed, this comment isn't quite right as we're using a fixed straight-line velocity.

But I think that's good - I don't think we want a fixed velocity in just the xz plane though. When you have to throw a present upwards more of the throw energy should go into the vertical speed and it should have a slower xz speed.

Comment (Jezzamonn) [reply static/scenes/railroad/js/present.js:9]: .... I am thinking about this more though, and I wonder if it makes sense to model things like that given that in practice, when you need to throw something further you do also need to throw things a lot faster too.

File: static/scenes/railroad/js/present.js [hunk 3]
```diff
@@ -50,39 +67,67 @@ class Present {
...
-        this.curve = new THREE.QuadraticBezierCurve3(this.startPosition.clone(), midpoint.clone(), this.targetPosition.clone());
+        // if we know gravity, the height delta, and how long the throw takes,
+        // we can choose a start y velocity that will last through the throw
+        this.velocityY = (this.targetPosition.y - this.startPosition.y) / this.durationOfThrow
+            - gravity * this.durationOfThrow / 2;
 
-        this.inFlight = true;
-    }
+        // the predicted linear Vy assuming no gravity
+        const initialVy = this.targetPosition.clone().sub(this.startPosition).normalize().y * linearThrowSpeed;
```
Comment (Jezzamonn) [inline static/scenes/railroad/js/present.js:92]: I added some logging to double check, and yeah this is the same as just using the `(this.targetPosition.y - this.startPosition.y) / this.durationOfThrow`, already calculated above.

```
:closure.js?module:8 Jez's eq = -6.345463343697578
:closure.js?module:9 initialVy = -6.3454633436975785
:closure.js?module:8 Jez's eq = 0.5083074855737352
:closure.js?module:9 initialVy = 0.5083074855737351
:closure.js?module:8 Jez's eq = -4.909785728905362
:closure.js?module:9 initialVy = -4.909785728905362
:closure.js?module:8 Jez's eq = 0.06645386784523435
:closure.js?module:9 initialVy = 0.06645386784523435
:closure.js?module:8 Jez's eq = -3.77779873567537
:closure.js?module:9 initialVy = -3.77779873567537
:closure.js?module:8 Jez's eq = -0.4738947319894831
:closure.js?module:9 initialVy = -0.4738947319894831
:closure.js?module:8 Jez's eq = 0.6739115767726478
:closure.js?module:9 initialVy = 0.6739115767726478
:closure.js?module:8 Jez's eq = 0.6797564477703062
:closure.js?module:9 initialVy = 0.6797564477703062
```

We can fix that up later though.

**Review by Jezzamonn — APPROVED** [review]
I think this feels pretty good. It's still not the nicest from a player POV when the presents leave the screen from being thrown too high vertically, but that doesn't happen as much now.

**Pull request merged**